torch/optim/_muon.py

```diff
@@ -0,0 +1,276 @@
+        X = a * X + B @ X
+
+    if grad.size(0) > grad.size(1):
+        X = X.T
+    return X
+
+
+def default_adjust_lr(lr: float, param_shape: torch.Size) -> float:
+    """Default learning rate adjustment used by Muon. Method reported in the paper https://arxiv.org/pdf/2502.16982."""
+    A, B = param_shape[:2]
+    adjusted_ratio = 0.2 * math.sqrt(max(A, B))
+    return lr * adjusted_ratio
+
+
+class Muon(Optimizer):
+    """Implements the Muon optimizer.
+
+    This optimizer performs momentum SGD followed by an optional orthogonalization
+    step computed via a user provided callable.
+    """
+
+    def __init__(
+        self,
+        params: ParamsT,
+        lr: float = 1e-3,
+        weight_decay: float = 0.1,
+        momentum: float = 0.95,
+        nesterov: bool = True,
+        *,
+        msign_fn: MsignFn = zeropower_via_newtonschulz,
+        msign_fn_config: BaseMsignFnConfig = NewtonSchulzConfig(),
+        adjust_lr_fn: AdjustLrFn = default_adjust_lr,
+    ) -> None:
+        if isinstance(lr, Tensor) and lr.numel() != 1:
+            raise ValueError("Tensor lr must be 1-element")
+        if not 0.0 <= lr:
+            raise ValueError(f"Learning rate should be >= 0 but is: {lr}")
+        if not 0.0 <= momentum:
+            raise ValueError(f"momentum should be >= 0 but is: {momentum}")
+        if not 0.0 <= weight_decay:
+            raise ValueError(f"weight decay should be >= 0 but is: {weight_decay}")
+
+        defaults = {
+            "lr": lr,
+            "weight_decay": weight_decay,
+            "momentum": momentum,
+            "nesterov": nesterov,
+        }
+        super().__init__(params, defaults)
+
+        for group in self.param_groups:
+            for p in group["params"]:
+                if p.ndim != 2:
+                    raise ValueError(
+                        f"Muon only supports 2D parameters where as the parameter has size: {p.size()}"
+                    )
+
+        self._msign_fn = msign_fn
+        self._msign_fn_config = msign_fn_config
+        self._adjust_lr_fn = adjust_lr_fn
+
+    def _init_group(
+        self,
+        group: MutableMapping,
+        params_with_grad: list[Tensor],
+        grads: list[Tensor],
+        muon_momentum_bufs: list[Tensor],
+    ):
+        for p in group["params"]:
+            if p.grad is None:
+                continue
+
+            if torch.is_complex(p):
+                raise RuntimeError("Muon does not support complex parameters")
+            if p.grad.is_sparse:
+                raise RuntimeError("Muon does not support sparse gradients")
+
+            params_with_grad.append(p)
+            grads.append(p.grad)
+
+            state = self.state[p]
+
+            buf = state.get("momentum_buffer")
+            if buf is None:
+                buf = torch.zeros_like(p.grad, memory_format=torch.preserve_format)
+                state["momentum_buffer"] = buf
+            muon_momentum_bufs.append(buf)
```

Comment:
```suggestion
            if state.get("momentum_buffer") is None:
                state.get("momentum_buffer") = torch.zeros_like(p.grad, memory_format=torch.preserve_format)
            muon_momentum_bufs.append(state.get("momentum_buffer"))
```
no need for buf intermediate, right

Reply:
just trying to reduce the number of times we mention "momentum_buffer". also, `state.get("momentum_buffer") = ` seems not right?

nonetheless, updated the code to

```
            if "momentum_buffer" not in state:
                state["momentum_buffer"] = torch.zeros_like(
                    p.grad, memory_format=torch.preserve_format
                )
            muon_momentum_bufs.append(state["momentum_buffer"])
```